### PR TITLE
Comprehensive fixes for deployment image triggering

### PR DESCRIPTION
**Pull request by ironcladlou**
The current deployment image triggering logic is not accounting for various edge cases related to image IDs (or lack thereof). Define all the triggering scenarios in the comprehensive unit test and plug the holes in the triggering code so that triggers work consistently.

One known side effect of the current flawed code is mis-firing triggers which update the container to a nil image ID, causing races described in #1443.

Base all comparisons on image ID, and add new state to image change trigger params to store the last image ID used for triggering.

### CODE REVIEW

**Comment by ironcladlou** [comment]
@smarterclayton @ncdc @pmorie PTAL.

**Comment by ncdc** [comment]
I'm wondering if we could just call ParseDockerImageReference for both container.Image and latest.DockerImageReference and then compare the 2?

**Comment by ironcladlou** [comment]
>I'm wondering if we could just call ParseDockerImageReference for both container.Image and latest.DockerImageReference and then compare the 2?

If `ImageRepository.DockerImageReference` *always* eventually contains the latest state (including image ID), I don't see why not. But that's not saying much, since I'm still learning all the IR mechanics on the fly. :)

**Comment by ironcladlou** [comment]
Talked with @ncdc, there's mork work to do here still. Hashing out unit tests currently.

**Comment by ncdc** [comment]
It's not `ImageRepository.DockerImageReference` - it's the `DockerImageReference`s created by parsing `container.Image` and `latest.DockerImageReference`. Very confusing, I know :-).

`status.tags.items[n].dockerImageReference` will always be a pull spec. Sometimes it will just be a pull spec to a tag, sometimes it will be a pull spec to a "pull by id" tag, and once we have v2 registries and Docker 1.6, it will be a pull spec to an image by its digest, for true pull by id semantics.

**Comment by ironcladlou** [comment]
@ncdc Do you think the new test support in https://github.com/ironcladlou/origin/commit/face43227f1bd05d07d33e81a8b69352be89a11f sufficient to express our scenarios?

**Comment by ncdc** [comment]
@ironcladlou I think so

**Comment by smarterclayton** [comment]
Status on this?

**Comment by ironcladlou** [comment]
Paused it while @ncdc was out to make progress on deployment hooks, will be revisiting it this afternoon.

**Comment by ironcladlou** [comment]
@ncdc helped me hash this out. I'm going to make a stateless utility method in the `pkg/image/api` which will be something like:

```go
func TagDiffers(tag string, pullspec string, repo ImageRepository) bool
```

The table tests in pkg/deploy/controller/imagechange/controller_test.go#TestHandle_matchScenarios` are currently conflating tests for the diff logic and the controller's core triggering logic. I'll use them as a guide to make tests for `TagDiffers` and simplify the controller test.

cc @bparees for builds.

**Comment by ironcladlou** [comment]
Nevermind, the proposed solution still isn't right. We're working on it.

**Comment by ironcladlou** [comment]
Latest branch has big refactorings to the API, image trigger controller, and generator to implement the correct behavior. The e2e tests are already working, but I need to go back and re-do the unit tests.

**Comment by ironcladlou** [comment]
Okay, @ncdc @smarterclayton and @pmorie.

I had to rewrite the deployment image change trigger controller, the generator, and all the related tests. I also had to make an additive/backwards-compatible API change to add a status struct to `DeploymentTriggerImageChangeParams` so we can correctly track the last triggered image ID.

The changes are significant, but as far as I know, the behavior is actually correct now in terms of ImageRepository interactions. Hopefully you all can do some auditing.

In the meantime, [test].

**Comment by ironcladlou** [comment]
I was also able to remove a metric ton of test cruft during the course of this by consolidating test fixtures and removing dead tests.

**Comment by smarterclayton** [comment]
> On Mar 30, 2015, at 3:40 PM, Dan Mace <notifications@github.com> wrote:
> 
> Okay, @ncdc @smarterclayton and @pmorie.
> 
> I had to rewrite the deployment image change trigger controller, the generator, and all the related tests. I also had to make an additive/backwards-compatible API change to add a status struct to DeploymentTriggerImageChangeParams so we can correctly track the last triggered image ID.
> 
So I have a question... Why is the pull spec in the pod template not sufficient?  Why do we need this other field?

**Comment by ncdc** [comment]
> Why is the pull spec in the pod template not sufficient?  Why do we need this other field?

We need the image id, which won't be in the pull spec if the registry it comes from doesn't support pull by id.

**Comment by smarterclayton** [comment]
Why do we need the image id?  If the pull spec doesn't change you can't guarantee you'll get a new image (although you can guarantee you'll get a new pod, you can't force the node to pull a new image).

----- Original Message -----

**Comment by ncdc** [comment]
True, but if the ImagePullPolicy is the default, it will always pull. This is the easiest way to be consistent for all our image change trigger semantics. The code was getting too ugly and too hard to reason through. Going forward in my IS PR, the TagEvents in Status.Tags[tag] must have both DockerImageReference and Image (id) to help enable this. Dan and I spent multiple hours talking through this on IRC and the phone, and this way is easy to understand and follow.

**Comment by ironcladlou** [comment]
@smarterclayton In [an earlier IRC conversation](https://botbot.me/freenode/openshift-dev/2015-03-24/?msg=34919505&page=4) you brought up "ID change with no pullspec change" use case (which this is all now based around). Did Andy and I misinterpret you originally or did you have some new thoughts on it?

**Comment by ironcladlou** [comment]
To be clear: I too am advocating for triggering exclusively on image ID.

File: pkg/deploy/api/types.go [hunk 1]
```diff
@@ -182,6 +182,14 @@ type DeploymentTriggerImageChangeParams struct {
 	From kapi.ObjectReference `json:"from"`
 	// Tag is the name of an image repository tag to watch for changes.
 	Tag string `json:"tag,omitempty"`
+	// Status contains information about the last trigger event.
+	Status DeploymentTriggerImageChangeParamsStatus `json:"status"`
```
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: Why is this different from BuildConfigTriggers?

**Comment by smarterclayton** [comment]
We have to trigger on pull spec.  It's less clear to me that we trigger when pull-spec doesn't change, although if you've done the work for it and it doesn't dramatically increase complexity I'm ok with it.

----- Original Message -----

**Comment by ncdc** [comment]
If we trigger on image id it's the same net effect as triggering on pull spec.

File: pkg/deploy/api/types.go (hunk 1 above)
Comment (ironcladlou) [inline pkg/deploy/api/types.go]: Isn't BuildConfigTriggers the odd one out here? It has fields that are spec-like configuration supplied by user, and then also a status-like LastVersion set by the system which isn't meant to be set by users. I felt like I was correcting that here by encapsulating the read-only system-set fields into status.

**Comment by ncdc** [comment]
Also, if you want to trigger on anything, you have to have an ImageStream, and with the poller importing images and their IDs, this enables us to trigger solely on image id.

File: pkg/deploy/api/types.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: Status and Spec are root level concepts.  They don't really apply deeper in the tree right now.

----- Original Message -----

**Comment by smarterclayton** [comment]
It does... today... but even in those cases, it's the pull spec changing that causes the pull, not anything else.  Triggering a deployment when the pull spec hasn't changed is questionable at best unless you combine it with an image policy.

----- Original Message -----

**Comment by ncdc** [comment]
I think we're going to end up with users scratching their heads if we don't trigger solely on image id. Imagine Joe User saying this:

> I push a new image to my repo, `joe/awesome:latest`. This means it should kick off a new deployment.

Joe's registry/repo doesn't support pull by id, but that shouldn't be a reason to penalize him by not deploying, should it?

**Comment by ncdc** [comment]
Also FYI in my ImageStream PR, I am removing support for setting a spec tag to an arbitrary pull spec or image ID that we don't have in status.tags. That was part of what was making the code so hard to follow. We'll have to allow users to update an external repo's status tags before you can set a spec tag or trigger off of a tag change in an external repo, either by polling periodically or giving them a manual means of updating the status.tags.

**Comment by smarterclayton** [comment]
In that case image id isn't helping, you just need a way to trigger the deployment.  The important distinction here is that *only* a change to the image pull spec guarantees a deployment.  If we import his registry but don't change the tag, when we trigger a new deployment we're not promising he's going to get a new image.  He has to rev the tag himself.

----- Original Message -----

**Comment by smarterclayton** [comment]
Sure, that's expected.

----- Original Message -----

**Comment by ncdc** [comment]
> In that case image id isn't helping, you just need a way to trigger the deployment.  The important distinction here is that *only* a change to the image pull spec guarantees a deployment. If we import his registry but don't change the tag, when we trigger a new deployment we're not promising he's going to get a new image.

Agreed, unless the pull policy is set correctly, but we can't easily guarantee that.

> He has to rev the tag himself.

From what to what? If the current TagEvent for latest is `DockerImageReference=joe/awesome:latest, Image=abcd1234` and he pushes a new image, the only thing that changes in the new TagEvent is `Image`. The pull spec, `DockerImageReference`, hasn't changed.

**Comment by smarterclayton** [comment]
He must physically change the tag on the docker end.  I'm provisionally ok with someone setting PullAlways (or maybe we need one which is PullAtLeastOnce) after the trigger a deploy but i'm not sure we require this change in order to reflect that.

----- Original Message -----

**Comment by ncdc** [comment]
> He must physically change the tag on the docker end

Is this what you mean?

1. `docker build -t joe/awesome:revision2 .`
2. `docker push joe/awesome:revision2`
3. push button to manually sync status.tags
4. update ImageStream, set spec.tags[latest] = ImageStreamTag(awesome:revision2)

at this point, with a trigger watching ImageStreamTag(awesome:latest), the pull spec for status.tags[latest].Items[0].DockerImageReference is now `joe/awesome:revision2`, which is different from before (`joe/awesome:latest`), thus triggering a deployment.

Isn't it a whole lot easier just to trigger on image id? :smile: We can always educate users to set the pull policy to always in this situation.

**Comment by smarterclayton** [comment]
----- Original Message -----
> > He must physically change the tag on the docker end
> 
> Is this what you mean?
> 
> 1. `docker build -t joe/awesome:revision2 .`
> 2. `docker push joe/awesome:revision2`
> 3. push button to manually sync status.tags
> 4. update ImageStream, set spec.tags[latest] =
> ImageStreamTag(awesome:revision2)
> 
> at this point, with a trigger watching ImageStreamTag(awesome:latest), the
> pull spec for status.tags[latest].Items[0].DockerImageReference is now
> `joe/awesome:revision2`, which is different from before
> (`joe/awesome:latest`), thus triggering a deployment.
> 
> Isn't it a whole lot easier just to trigger on image id? :smile: We can
> always educate users to set the pull policy to always in this situation.

You don't need to physically store the image ID to trigger a build - that happens automatically as long as you can make at least one edit to the deployment config (or just to the latest version, but that's a bit tacky right now).

File: pkg/deploy/api/types.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/deploy/api/types.go]: Flatten this and add a TODO in BuildConfig to rename to LastTriggeredImage.  And it's not really an ID, it's the image's  name or unique identifier (but not uid).

**Comment by ncdc** [comment]
So are you saying then that all that needs to happen is:

1. `docker build -t joe/awesome:latest`
2. `docker push joe/awesome:latest`
3. update deployment config
4. profit?

File: pkg/deploy/controller/imagechange/controller.go [hunk 2]
```diff
@@ -26,72 +24,64 @@ func (e fatalError) Error() string { return "fatal error handling imageRepositor
...
+			// Find the latest tag event for the trigger tag
+			latestEvent, err := imageapi.LatestTaggedImage(imageRepo, params.Tag)
+			if err != nil {
+				glog.V(4).Infof("Couldn't find latest tag event for tag %s in imageRepo %s: %s", params.Tag, labelForRepo(imageRepo), err)
+				continue
+			}
+
+			// If the tag event doesn't have an image ID yet, we can't determine
+			// whether there was really a change
+			if len(latestEvent.Image) == 0 {
```
Comment (smarterclayton) [inline pkg/deploy/controller/imagechange/controller.go]: Why not?  Why can't you just use the pull spec like Builds?

Comment (ncdc) [inline pkg/deploy/controller/imagechange/controller.go]: This actually won't be an issue once my ImageStream PR goes in - Image won't ever be empty

Comment (smarterclayton) [inline pkg/deploy/controller/imagechange/controller.go]: What if I want to tag an arbitrary docker repository on the docker hub?

----- Original Message -----

Comment (ncdc) [inline pkg/deploy/controller/imagechange/controller.go]: You have to create an ImageStream for it

Comment (smarterclayton) [inline pkg/deploy/controller/imagechange/controller.go]: ----- Original Message -----
> > -				}
> > -				if latest.Image != containerImageID {
> > -					glog.V(4).Infof("Container %s for config %s: image id changed from %q
> > to %q; regenerating config", container.Name, labelFor(config),
> > containerImageID, latest.Image)
> > -					configsToGenerate = append(configsToGenerate, config)
> > -					firedTriggersForConfig[config.Name] =
> > append(firedTriggersForConfig[config.Name], params)
> > -				}
> > +			// Find the latest tag event for the trigger tag
> > +			latestEvent, err := imageapi.LatestTaggedImage(imageRepo, params.Tag)
> > +			if err != nil {
> > +				glog.V(4).Infof("Couldn't find latest tag event for tag %s in
> > imageRepo %s: %s", params.Tag, labelForRepo(imageRepo), err)
> > +				continue
> > +			}
> > +
> > +			// If the tag event doesn't have an image ID yet, we can't determine
> > +			// whether there was really a change
> > +			if len(latestEvent.Image) == 0 {
> 
> You have to create an ImageStream for it

I don't think we are on the same page about that.

File: pkg/deploy/api/types.go (hunk 1 above)
Comment (ironcladlou) [inline pkg/deploy/api/types.go]: Roger, will change it. Thanks for the clarification.

File: pkg/deploy/controller/imagechange/controller.go (hunk 2 above)
Comment (ncdc) [inline pkg/deploy/controller/imagechange/controller.go]: @ironcladlou I talked to @smarterclayton and agree that a user shouldn't have to go through the hoops of creating an ImageStream that points to an external Docker image repository just to be able to deploy something. Let's talk tomorrow about how to trigger off of pull spec instead of image (it shouldn't be too much of a change).

Comment (ironcladlou) [inline pkg/deploy/controller/imagechange/controller.go]: Sounds good, will catch up with you and figure out how to refactor.

**Comment by smarterclayton** [comment]
Is this going to fix the integration test race?

**Comment by ironcladlou** [comment]
I re-centered the logic around `TagEvent.Status.DockerImageReference` rather than `.Image`. PTAL @smarterclayton @ncdc

**Comment by ironcladlou** [comment]
Not sure yet exactly what failed in that jenkins run, so I want to run this through [test] again.

**Comment by ironcladlou** [comment]
Still not sure what's going on with jenkins. Unit, integration, and e2e tests are passing.

**Comment by ironcladlou** [comment]
https://github.com/openshift/origin/pull/1550/files is what is up with Jenkins. @smarterclayton PTAL, I think your comments are addressed.

File: pkg/deploy/controller/imagechange/controller.go [hunk 3]
```diff
@@ -26,72 +24,55 @@ func (e fatalError) Error() string { return "fatal error handling imageRepositor
...
-					// For v1 images, the container image's tag name is by convention the same as the image ID it references
-					containerImageID = ref.Tag
-				}
-				if latest.Image != containerImageID {
-					glog.V(4).Infof("Container %s for config %s: image id changed from %q to %q; regenerating config", container.Name, labelFor(config), containerImageID, latest.Image)
-					configsToGenerate = append(configsToGenerate, config)
-					firedTriggersForConfig[config.Name] = append(firedTriggersForConfig[config.Name], params)
-				}
+			// Ensure a change occured
+			if latestEvent.DockerImageReference != params.LastTriggeredImage {
```
Comment (smarterclayton) [inline pkg/deploy/controller/imagechange/controller.go]: If latestEvent.DockerImageReference is empty you shouldn't trigger an update.

Comment (ncdc) [inline pkg/deploy/controller/imagechange/controller.go]: It will never be empty

Comment (ironcladlou) [inline pkg/deploy/controller/imagechange/controller.go]: This is now accounted for.

**Comment by smarterclayton** [comment]
LGTM, squash

**Comment by ironcladlou** [comment]
Squashed

**Comment by smarterclayton** [comment]
[merge]
